### PR TITLE
fixed object orientation

### PR DESCRIPTION
**Pull request by nlyubova**
detection of several objects at the same time

### CODE REVIEW

File: src/linemod_icp.cpp [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+  }
+
+  float ratio = 0.0f;
+  if ((src_ref.cols > 0) && (src_ref.rows > 0))
+    ratio = float(px_missing) / float(src_ref.cols * src_ref.rows);
+  return ratio;
+}
+
+/** Adjusts the bounding box to the image size */
+bool boundingboxAdjust(cv::Rect_<int> &rect, const int &w, const int &h)
```
Comment (nlyubova) [inline src/linemod_icp.cpp]: done!

File: src/linemod_icp.cpp [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+      if (dist < dist_expected)
+        ++nbr_inliers;
+    }
+
+    ++counter;
+  }
+
+  if (counter > 0)
+  {
+    dist_mean /= (float)counter;
```
Comment (nlyubova) [inline src/linemod_icp.cpp]: true

File: src/linemod_detect.cpp [hunk 3]
```diff
@@ -162,24 +319,149 @@ struct Detector: public object_recognition_core::db::bases::ModelReaderBase {
...
+      std::vector<cv::Vec3f> pts_real_model_temp;
+      std::vector<cv::Vec3f> pts_real_ref_temp;
+      float px_ratio_missing = matToVec(depth_real_model, depth_real_ref, pts_real_model_temp, pts_real_ref_temp);
+      if (px_ratio_missing > *px_match_min_)
+        continue;
+
+      //perform the first approximate ICP
+      float px_ratio_match_inliers = 0.0f;
+      float icp_dist = icpCloudToCloud(pts_real_ref_temp, pts_real_model_temp, R_real_icp, T_real_icp, px_ratio_match_inliers, 1);
+      //reject the match if the icp disntance if too big
```
Comment (nlyubova) [inline src/linemod_detect.cpp]: true

File: src/linemod_icp.cpp [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+{
+  centroid = cv::Vec3f(0.0f, 0.0f, 0.0f);
+  size_t n_points = 0;
+  for (std::vector<cv::Vec3f>::const_iterator it = pts.begin(); it != pts.end(); it++) {
+    if (!cv::checkRange(*it))
+      continue;
+    centroid += (*it);
+    ++n_points;
+  }
+  centroid(0) /= n_points;
```
Comment (nlyubova) [inline src/linemod_icp.cpp]: done

File: src/linemod_pointcloud.h [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#ifndef LINEMOD_POINTCLOUD_H
+#define LINEMOD_POINTCLOUD_H
+
+#include <ros/ros.h>
+#include "sensor_msgs/PointCloud2.h"
```
Comment (nlyubova) [inline src/linemod_pointcloud.h]: done

File: src/linemod_train.cpp [hunk 6]
```diff
@@ -50,7 +49,6 @@
 #include <object_recognition_core/db/document.h>
 #include <object_recognition_core/db/model_utils.h>
 
-#define USE_GLUT 1
```
Comment (nlyubova) [inline src/linemod_train.cpp]: I could keep it. 
:) I was also wondering why did you add it by default?

Comment (vrabaud) [inline src/linemod_train.cpp]: it's because GLUT worked better for me. But it actually requires an X server. If OSMesa is stable enough and works for you, then let's keep it as is for now, no change. Thx.

File: src/linemod_detect.cpp [hunk 7]
```diff
@@ -162,24 +319,149 @@ struct Detector: public object_recognition_core::db::bases::ModelReaderBase {
...
+
+        //perform the final precise icp
+        float icp_px_match = 0.0f;
+        float icp_dist = icpCloudToCloud(o_match->pts_ref, o_match->pts_model, o_match->r, o_match->t, icp_px_match, 0);
+
+        if (*verbose_)
+          std::cout << o_match->match_class << " " << o_match->match_sim << " icp " << icp_dist << ", ";
+
+        //additional depth check
+        icp_dist = 0.007f;
```
Comment (nlyubova) [inline src/linemod_detect.cpp]: the same as sensor data, since it is the distance between points.
this distance is used to compute the ratio of inliers (points laying within a certain distance between two point clouds; In the paper, this distance is equal to the mean distance*3 and at the beginning, they use all points)

Comment (vrabaud) [inline src/linemod_detect.cpp]: please just write it as a comment above or something

File: src/linemod_train.cpp [hunk 8]
```diff
@@ -130,46 +134,59 @@ namespace ecto_linemod
...
-      cv::Mat image, depth, mask;
-      cv::Rect rect;
-      cv::Mat_<unsigned short> depth_short;
-      cv::Matx33d R;
-      cv::Vec3d T;
-      for (size_t i = 0; !renderer_iterator.isDone(); ++i, ++renderer_iterator)
-      {
+    RendererIterator renderer_iterator = RendererIterator(&renderer, 150);
+    //set the RendererIterator parameters
+    renderer_iterator.angle_step_ = 10;
```
Comment (nlyubova) [inline src/linemod_train.cpp]: ok, should i set them equal to all objects or different for each object?

File: src/linemod_detect.cpp (hunk 7 above)
Comment (nlyubova) [inline src/linemod_detect.cpp]: done

File: src/linemod_train.cpp (hunk 6 above)
Comment (nlyubova) [inline src/linemod_train.cpp]: ok

File: src/linemod_train.cpp (hunk 8 above)
Comment (vrabaud) [inline src/linemod_train.cpp]: for now, let them as is in the code. But they should be set per object as, in the future, they will be computed per object.
